tests/src/Integration/Action/UserRoleAddTest.php

```diff
@@ -0,0 +1,214 @@
+   *
+   * @covers ::execute
+   */
+  public function testAddThreeRoles() {
+    // Set-up a mock user.
+    $account = $this->getMockedUser();
+    // Mock hasRole.
+    $account->expects($this->exactly(3))
+      ->method('hasRole')
+      ->with($this->logicalOr(
+        $this->equalTo('manager'),
+        $this->equalTo('editor'),
+        $this->equalTo('administrator')
+      ))
+      ->will($this->returnValue(FALSE));
+    // Mock addRole.
+    $account->expects($this->exactly(3))
+      ->method('addRole')
+      ->with($this->logicalOr(
+        $this->equalTo('manager'),
+        $this->equalTo('editor'),
+        $this->equalTo('administrator')
+      ));
+
+    // Mock user roles.
+    $manager = $this->getMockedUserRole('manager');
+    $editor = $this->getMockedUserRole('editor');
+    $administrator = $this->getMockedUserRole('administrator');
+
+    // Test adding of three roles role.
+    $this->action
+      ->setContextValue('user', $account)
+      ->setContextValue('roles', [$manager, $editor, $administrator])
+      ->execute();
+  }
+
+  /**
+   * Tests adding of existing role to user.
+   *
+   * @covers ::execute
+   */
+  public function testAddExistingRole() {
+    // Set-up a mock user with role 'administrator'.
+    $account = $this->getMockedUser();
+    $account->expects($this->once())
+      ->method('hasRole')
+      ->with($this->equalTo('administrator'))
+      ->will($this->returnValue(TRUE));
+
+    // We do not expect a call of the 'addRole' method.
+    $account->expects($this->never())
+      ->method('addRole');
+
+    // Mock the 'administrator' user role.
+    $administrator = $this->getMockedUserRole('administrator');
+
+    // Test adding one role.
+    $this->action
+      ->setContextValue('user', $account)
+      ->setContextValue('roles', [$administrator])
+      ->execute();
+  }
+
+  /**
+   * Tests adding of one existing and one nonexistent role to user.
+   *
+   * @covers ::execute
+   */
+  public function testAddExistingAndNonexistentRole() {
+    // Set-up a mock user with role 'administrator' but without 'editor'.
+    $account = $this->getMockedUser();
+    $account->expects($this->exactly(2))
+      ->method('hasRole')
+      ->with($this->logicalOr(
+        $this->equalTo('editor'),
+        $this->equalTo('administrator')
+      ))
+      ->will($this->returnCallback(
+        function($rid) {
+          if ($rid == 'administrator') {
+            return TRUE;
+          }
+          return FALSE;
+        }
+      ));
+
+    // We expect only one call of the 'addRole' method.
+    $account->expects($this->once())
+      ->method('addRole');
+
+    // Mock user roles.
+    $editor = $this->getMockedUserRole('editor');
+    $administrator = $this->getMockedUserRole('administrator');
+
+    // Test adding one role.
+    $this->action
+      ->setContextValue('user', $account)
+      ->setContextValue('roles', [$administrator, $editor])
+      ->execute();
```

Comment:
After that we need to call autoSaveContext() on the plugin and make sure that it returns ['user'].